internal/states/checks.go

```diff
@@ -0,0 +1,182 @@
+// checks in the configuration. For our UI reporting purposes, it is entire
+// objects that pass or fail based on their declared checks; the individual
+// checks have no durable identity between runs, and so are only a language
+// design convenience to help authors describe various independent conditions
+// with different failure messages each.
+//
+// CheckResults should typically be considered immutable once constructed:
+// instead of updating it in-place,instead construct an entirely new
+// CheckResults object based on a fresh checks.State.
+type CheckResults struct {
+	// ConfigResults has all of the individual check results grouped by the
+	// configuration object they relate to.
+	//
+	// The top-level map here will always have a key for every configuration
+	// object that includes checks at the time of evaluating the results,
+	// even if there turned out to be no instances of that object and
+	// therefore no individual check results.
+	ConfigResults addrs.Map[addrs.ConfigCheckable, *CheckResultAggregate]
+}
+
+// CheckResultAggregate represents both the overall result for a particular
+// configured object that has checks and the individual checkable objects
+// it declared, if any.
+type CheckResultAggregate struct {
+	// Status is the aggregate status across all objects.
+	//
+	// Sometimes an error or check failure during planning will prevent
+	// Terraform Core from even determining the individual checkable objects
+	// associated with a downstream configuration object, and that situation is
+	// described here by this Status being checks.StatusUnknown and there being
+	// no elements in the ObjectResults field.
+	//
+	// That's different than Terraform Core explicitly reporting that there are
+	// no instances of the config object (e.g. a resource with count = 0),
+	// which leads to the aggregate status being checks.StatusPass while
+	// ObjectResults is still empty.
+	Status checks.Status
+
+	ObjectResults addrs.Map[addrs.Checkable, *CheckResultObject]
+}
+
+// CheckResultObject is the check status for a single checkable object.
+//
+// This aggregates together all of the checks associated with a particular
+// object into a single pass/fail/error/unknown result, because checkable
+// objects have durable addresses that can survive between runs, but their
+// individual checks do not. (Module authors are free to reorder their checks
+// for a particular object in the configuration with no change in meaning.)
+type CheckResultObject struct {
+	// Status is the check status of the checkable object, derived from the
+	// results of all of its individual checks.
+	Status checks.Status
+
+	// FailureMessages is an optional set of module-author-defined messages
+	// describing the problems that the checks detected, for objects whose
+	// status is checks.StatusFail.
+	//
+	// (checks.StatusError problems get reported as normal diagnostics during
+	// evaluation instead, and so will not appear here.)
+	FailureMessages []string
```

Comment:
A very wooly thought about this structure: in future we might want to add more failure information for an individual check, such as a label, or a condition expression representation. I think that would mean changing this into a set of objects, like `Failures []CheckResultFailure`.

I think that in our JSON serialization we already make space for this by creating an object per failure with a `"message"` key. Is it worth doing the same thing here for ease of migration alter, or is it too early?

Reply:
I did indeed intentionally make the JSON serialization have room for expansion, because that's the only part that will become a compatibility constraint moving forward.

Honestly if this had come about in a different order where I knew I'd be prioritizing the JSON output rather than the `terraform test` command I likely would've built this as a slice of some struct, but while I was revising to change the focus I had concluded that it wouldn't be a big deal to change this internal model later if we need to.

With that said, this also probably isn't a big deal to change proactively now. I think what I will do is land this PR with it implemented this way and then I'll make a smaller follow-up PR that deals with this refactoring which we can safely land at any point in the v1.3 maintenance period without changing any externally-visible behavior. (and similarly for your other feedback bits above, which I think are also worth exploring.)

Reply:
I did take a quick attempt at this just after merging and found that it's a little trickier than I hoped because it really ought to be extended in both the `checks` package and the `states` package in a similar way, and I want to spend a little more time pondering whether it's useful to share a struct type between both of those or whether it's worth keeping them separate, and other similar internal API design questions.

I do intend to pick this up again later but Friday-afternoon brain is insufficient for this sort of design work so I'll just keep it in mind for exploring in the near future. Thanks for pointing it out!